### ISSUE TITLE
CRASH with multiple native handlers

### ISSUE THREAD

**Issue by algrant-arm** [DynamoRIO/dynamorio#5898]
DynamoRIO 9.0.1 and trunk crash with an error about "multiple native handlers", in cases involving a sigaction after a vfork or clone:
```
#include <signal.h>
#include <unistd.h>
int main ()
{
  int r;
  static struct sigaction sa;
  sigaction (SIGUSR1, 0, &sa);
  r = vfork();
  if (r == 0) _exit(1);
  sigaction (SIGUSR1, &sa, 0);
  return 0;
}
```
$ DynamoRIO-AArch64-Linux-9.0.1/bin64/drrun -- ./a.out
<Application /home/agrant/gcscheck/a.out (9637). Cannot correctly handle received signal 11 in thread 9637: multiple native handlers.>
```
Signal 11 is SIGSEGV.
```
$ DynamoRIO-AArch64-Linux-9.0.1/bin64/drrun -debug -- ./a.out
<Starting application /home/agrant/gcscheck/a.out (9635)>
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/home/agrant/gcscheck/DynamoRIO-AArch64-Linux-9.0.1/lib64/debug/libdynamorio.so' 0x0000000071015e18
>
<(1+x) Handling our fault in a TRY at 0x000000007137726c>
<Application /home/agrant/gcscheck/a.out (9635). Cannot correctly handle received signal 17 in thread 9635: multiple native handlers.>
```
Signal 17 is SIGCHLD.

Looking at core/unix/signal.c, the comment around this failure refers to another comment:
```
... i#1921: To properly handle multiple separate signal handler thread groups
 * within this single DR process domain, we would need a list of these, along with
 * lists of threads within each.  Since that is very rare, and requires either
 * hardcoded maximums or complexities with exit-time heap, we currently do not
 * support that and die if asked to deliver a native signal in such circumstances.
```
But that doesn't seem relevant to this test case, which was extracted from posix_spawn() in glibc.

**Comment by derekbruening** [comment]
A "native signal" or "native handler" refers to a situation where DR does not control the thread: where it is "native" and not running managed in the code cache.  That should never happen with a plain execution of a target program: that only happens after an explicit detach or if certain options are passed.  Looking at the code, there seems to be a case of a not-yet-initialized thread where it falls back to execute_native_handler(): I would look at why it got there and why it is not initialized: presumably something related to the vfork.

**Comment by algrant-arm** [comment]
Having traced this some more, I think it might be a problem in TLS. The child thread terminates and calls tls_thread_free(). This sets *dr_tls_base_addr to TLS_SLOT_VAL_EXITED. But the parent, which is in the same address space, sees the same value of dr_tls_base_addr (this is presumably an error - different threads should see different values?). The parent now sees TLS_SLOT_VAL_EXITED, so is_thread_tls_initialized() returns false, attempts to get the context return NULL, and DynamoRIO thinks it's now a native thread, and then when the parent gets SIGCHILD it gets the multiple native handlers error.

**Comment by derekbruening** [comment]
Is the issue that the kernel does not create separate TLS (x86-64: new %fs; on aarch64: new tpidrurw or whichever reg it is) for a vfork child, unlike a clone child?  And DR doesn't handle the two using the same TLS?  I know DR has a bunch of code dealing with cleaning up after a vfork child does an execve; probably we have never seen a vfork exit before.

**Comment by algrant-arm** [comment]
Yes that does seem to be the issue. tpidr_el0 is inherited by the child, so it's sharing the parent's TLS, which it then clears up on its own exit (e.g. removing the dcontext) and then the parent gets confused when it resumes. I guess there needs to be some state in the (shared) dcontext to say "right now, we are in a vfork() child, with the parent suspended", and then if we see exit, we know it's from the child, so we avoid doing a thread_exit. We should unset the flag saying we're in a vfork() child. And possibly unset the multiple_native_handlers flag too (or better, make it an atomic counter and decrement it?).
